providers/implementations/kem/template_kem.c

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/crypto.h>
+#include <openssl/evp.h>
+#include <openssl/core_dispatch.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include <openssl/err.h>
+#include <openssl/proverr.h>
+#include "prov/provider_ctx.h"
+#include "prov/implementations.h"
+#include "prov/securitycheck.h"
+#include "prov/providercommon.h"
+
+extern const OSSL_DISPATCH ossl_template_asym_kem_functions[];
+
+#if defined(NDEBUG) || defined(OPENSSL_NO_STDIO)
+static void debug_print(char *fmt, ...)
+{
+}
+
+#else
+static void debug_print(char *fmt, ...)
+{
+    char out[1000];
+    va_list argptr;
+
+    va_start(argptr, fmt);
+    vsprintf(out, fmt, argptr);
+    va_end(argptr);
+    if (getenv("TEMPLATEKEM"))
+        fprintf(stderr, "TEMPLATE_KEM: %s", out);
+}
+#endif
+
+typedef struct {
+    OSSL_LIB_CTX *libctx;
+    /* some key struct */
+    int op;
+} PROV_TEMPLATE_CTX;
+
+static OSSL_FUNC_kem_newctx_fn template_newctx;
+static OSSL_FUNC_kem_encapsulate_init_fn template_encapsulate_init;
+static OSSL_FUNC_kem_encapsulate_fn template_encapsulate;
+static OSSL_FUNC_kem_decapsulate_init_fn template_decapsulate_init;
+static OSSL_FUNC_kem_decapsulate_fn template_decapsulate;
+static OSSL_FUNC_kem_freectx_fn template_freectx;
+static OSSL_FUNC_kem_set_ctx_params_fn template_set_ctx_params;
+
```

Comment:
```suggestion
static OSSL_FUNC_kem_settable_ctx_params_fn template_settable_ctx_params;
```

Reply:
Are the dup, get, and gettable functions needed?

Reply:
As far as I recall from doing `oqsprovider`, Yes. But as stated elsewhere, there is no definite list of functions that need to be supported for specific `openssl` operations to succeed. These skeletons are meant to enable TLS1.3 KEM operations only. If you'd have time to contribute, a matrix of "OpenSSL functionality" -> "required provider functions implemented" (and suitably extended skeletons, stating these dependencies) would be great to have. I think @levitte is looking into this, too.

Reply:
comments added.

Reply:
Looks ok to me... Can you add the suggestion above..